substrate/client/network/src/protocol/notifications/behaviour.rs

```diff
@@ -4564,4 +4565,310 @@ mod tests {
+			let event = match t {
+				0 => NotifsHandlerOut::OpenDesiredByRemote { protocol_index },
+				1 => NotifsHandlerOut::CloseDesired { protocol_index },
+				2 => NotifsHandlerOut::CloseResult { protocol_index },
+				3 => NotifsHandlerOut::OpenDesiredByRemote { protocol_index }, // TODO OpenResultOk
+				4 => NotifsHandlerOut::OpenResultErr { protocol_index },
+				5 => {
+					let b = Vec::<u8>::arbitrary(u)?;
+					let message = b[..].into();
+					NotifsHandlerOut::Notification { protocol_index, message }
+				},
+				_ => {
+					unreachable!("no such event type is possible")
+				},
+			};
+			let r = AFromHandler(pid, conn, event);
+			Ok(r)
+		}
+	}
+
+	#[derive(Arbitrary, Debug)]
+	enum FuzzAction {
+		FromSwarm(AFromSwarm),
+		FromHandler(AFromHandler),
+	}
+
+	#[test]
+	fn fuzz_states() {
+		fn fuzz_states_impl(u: &mut Unstructured<'_>) -> AResult<()> {
+			let actions = u.arbitrary::<Vec<FuzzAction>>()?;
+			let connected = ConnectedPoint::Listener {
+				local_addr: Multiaddr::empty(),
+				send_back_addr: Multiaddr::empty(),
+			};
+			let (mut notif, _peerset) = development_notifs();
+			assert_eq!(1, notif.notif_protocols.len());
+
+			for action in actions {
+				match action {
+					FuzzAction::FromSwarm(AFromSwarm(pid, e, n)) => {
+						// build event
+						let peer_id = pid.0;
+						let conn = ConnectionId::new_unchecked(n);
+						let event = match e {
+							FromSwarmE::ConnectionEstablished => {
+								let event = ConnectionEstablished {
+									peer_id: peer_id.clone(),
+									connection_id: conn,
+									endpoint: &connected,
+									failed_addresses: &[],
+									other_established: 0usize,
+								};
+								FromSwarm::ConnectionEstablished(event)
+							},
+							FromSwarmE::ConnectionClosed => {
+								let event = ConnectionClosed {
+									peer_id: peer_id.clone(),
+									connection_id: conn,
+									endpoint: &connected,
+									handler: NotifsHandler::new(
+										peer_id.clone(),
+										connected.clone(),
+										vec![],
+									),
+									remaining_established: 0usize,
+								};
+								FromSwarm::ConnectionClosed(event)
+							},
+							FromSwarmE::DialFailure => {
+								let event = DialFailure {
+									peer_id: Some(peer_id.clone()),
+									error: &libp2p::swarm::DialError::Banned,
+									connection_id: conn,
+								};
+								FromSwarm::DialFailure(event)
+							},
+						};
+						let entry_before = notif.peers.get(&(peer_id, 0.into())).cloned();
+						// precheck
+						match &event {
+							FromSwarm::ConnectionEstablished(_) => {
+                            },
+                            FromSwarm::ConnectionClosed(_) => {
+                                if matches!(entry_before, None) {
+									continue
+								}
+								if matches!(entry_before, Some(PeerState::Requested | PeerState::PendingRequest { .. } | PeerState::Backoff { .. } | PeerState::Poisoned)) {
```

Comment:
I'm not 100% sure about this logic so it would need a comment at least. Why is `ConnectionClosed` ignored if `PeerState` is `Requested` or `PendingRequest`?